### PR TITLE
Tools 2139 valgrind integration test

### PR DESCRIPTION
**Pull request by atefehmohseni**

### CODE REVIEW

File: test/integration/lib.py [hunk 1]
```diff
@@ -145,16 +152,32 @@ def temporary_path(extension):
 	GLOBALS["file_count"] += 1
 	return absolute_path(os.path.join(WORK_DIRECTORY, file_name))
 
-def run(command, *options, do_async=False, pipe_stdout=None, pipe_stdin=None, env={}):
+def run(command, *options, do_async=False, pipe_stdout=None, pipe_stdin=None, env={}, RUN_IN_DOCKER=False, USE_VALGRIND= False):
```
Comment (atefehmohseni) [inline test/integration/lib.py]: remove extra space in `USE_VALGRIND= False`

File: test/integration/run_backup.py [hunk 2]
```diff
@@ -168,3 +165,78 @@ def backup_and_restore(filler, preparer, checker, env={}, backup_opts=None,
...
+			res = lib.parse_val_logs("val_rs.log")
+		except:
+			as_srv.reset_aerospike_servers()
+			raise
+		as_srv.reset_aerospike_servers()
+		return res
+	print("cannot proceed restore tests with valgrind")
+	return False
+
+def run_backup_w_valgrind(filler, context={}, backup_options=None):
```
Comment (atefehmohseni) [inline test/integration/run_backup.py:166]: run against asbackup version that had memory leak issue

File: include/encode.h [hunk 3]
```diff
@@ -87,8 +87,8 @@ typedef enum {
 	PATH_TYPE_STRING,
 	// The path results in an integer.
 	PATH_TYPE_NUMERIC,
-	// The path results in a geojson value.
-	PATH_TYPE_GEOJSON
+	// The path results in a geo2sphere value.
```
Comment (atefehmohseni) [inline include/encode.h]: geo2dsphere typo (missed d) in the comment

File: src/utils.c [hunk 4]
```diff
@@ -1546,8 +1546,8 @@ parse_index_info(char *ns, char *index_str, index_param *index)
 				type = PATH_TYPE_NUMERIC;
 			} else if (strcasecmp(arg, "INT SIGNED") == 0) {
 				type = PATH_TYPE_NUMERIC;
-			} else if (strcasecmp(arg, "GEOJSON") == 0) {
-				type = PATH_TYPE_GEOJSON;
+			} else if (strcasecmp(arg, "GEO2DSPHERE") == 0) {
```
Comment (atefehmohseni) [inline src/utils.c]: both `GEOJSON` and `GEO2DSPHERE` should be accepted

File: test/integration/aerospike_servers.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
 
 WORK_DIRECTORY = lib.WORK_DIRECTORY
 
-SERVER_IMAGE = "aerospike/aerospike-server:5.7.0.17"
+SERVER_IMAGE = "aerospike/aerospike-server"
```
Comment (atefehmohseni) [inline test/integration/aerospike_servers.py:25]: check if `test_geo_index` was failed on previous versions too, and why?

File: test/integration/test_in_docker.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# coding=UTF-8
+
+"""
+Tests the representation of asbackup/asrestore output runned with valgrind inside a docker container
```
Comment (atefehmohseni) [inline test/integration/test_in_docker.py]: explain that this script doesnt run among other integration tests

Comment (atefehmohseni) [reply test/integration/test_in_docker.py]: script `test_in_docker` and related functions were removed from the integration tests directory. It was confusing and not necessary given `test_w_valgrind` already runs `asbackup/asrestore` against Valgrind among other tests before each build on Github.

File: src/utils.c (hunk 4 above)
Comment (atefehmohseni) [reply src/utils.c]: done

File: include/encode.h (hunk 3 above)
Comment (atefehmohseni) [reply include/encode.h]: done

File: test/integration/lib.py (hunk 1 above)
Comment (atefehmohseni) [reply test/integration/lib.py]: done

File: test/integration/run_backup.py (hunk 2 above)
Comment (dwelch-spike) [reply test/integration/run_backup.py:166]: Did you verify that this catches the issue with the older version?

File: test/integration/lib.py [hunk 7]
```diff
@@ -145,16 +145,20 @@ def temporary_path(extension):
...
 	directory = absolute_path("../..")
+	
 	command = [os.path.join("test_target", command)] + list(options)
 
+	val_args = "--track-fds=yes --leak-check=full --track-origins=yes --show-reachable=yes --suppressions={0}".\
+			format(absolute_path(VAL_SUP_FILE))
+	
 	if USE_VALGRIND:
-		command = ["./val.sh"] + command
+		command = "valgrind {0} -v".format(val_args).split() + command
```
Comment (dwelch-spike) [inline test/integration/lib.py:161]: why do we need to use val_args and not ./val.sh?

File: test/integration/run_backup.py [hunk 8]
```diff
@@ -168,3 +163,36 @@ def backup_and_restore(filler, preparer, checker, env={}, backup_opts=None,
...
+	try:
+		with open("val_bc.log", "w") as pipe_stdout:
+			lib.run("asbackup", *backup_options, do_async=False,
+				pipe_stdout=pipe_stdout, env={}, USE_VALGRIND=True)
+		res = lib.parse_val_logs("val_bc.log")
+	except:
+		as_srv.reset_aerospike_servers()
+		raise
+	as_srv.reset_aerospike_servers()
+	return res
```
Comment (dwelch-spike) [inline test/integration/run_backup.py:181]: Suggestion: you could use `try`, `except`, `finally `here, so that `as_srv.reset_aerospike_servers()` only appears in the finally block.

File: test/integration/run_backup.py [hunk 9]
```diff
@@ -168,3 +163,36 @@ def backup_and_restore(filler, preparer, checker, env={}, backup_opts=None,
...
+	try:
+		with open("val_rs.log", "w") as pipe_stdout:
+			lib.run("asrestore", *restore_options, do_async=False,
+				pipe_stdout=pipe_stdout, env={}, USE_VALGRIND=True)
+		res = lib.parse_val_logs("val_rs.log")
+	except:
+		as_srv.reset_aerospike_servers()
+		raise
+	as_srv.reset_aerospike_servers()
+	return res
```
Comment (dwelch-spike) [inline test/integration/run_backup.py:197]: same as above here

File: test/integration/test_w_valgrind.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    """
+	Tests restore to dir with batch write disabled running by valgrind
+	"""
+    restore_options = get_basic_restore_options()
+    assert run_restore_w_valgrind(*restore_options, "--disable-batch-writes") == True, "Restore test with valgrind failed, cmd options {0}".format(restore_options)
+
+def test_restore_batch_writes_to_dir():
+    """
+	Tests restore to dir with batch write enabled running by valgrind
+	"""
```
Comment (dwelch-spike) [inline test/integration/test_w_valgrind.py]: nit: the formatting for the quotes looks off. Can you move the bottom inline with the top?

File: val.sh [hunk 11]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 
-valgrind --leak-resolution=high --track-fds=yes --leak-check=full --track-origins=yes --show-reachable=yes --suppressions=val.supp ${*}
+/usr/bin/valgrind --leak-resolution=high --track-fds=yes --leak-check=full --track-origins=yes --show-reachable=yes --suppressions=val.supp ${*}
```
Comment (dwelch-spike) [inline val.sh]: Why was this changed? I think we should leave this without a complete path. Valgrind should be in a user's path and if it isn't they can change the command accordingly.

File: test/integration/test_w_valgrind.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
+# coding=UTF-8
+
+"""
+Tests the representation of asbackup/asrestore output runned with valgrind
```
Comment (dwelch-spike) [inline test/integration/test_w_valgrind.py]: Suggestion: maybe change the comment to something like "Tests basic asbackup/asrestore commands with valgrind for memory leaks/errors."

**Review by dwelch-spike — COMMENTED** [review]
Just have a few nitpicks and question. Code looks good otherwise. :)

File: test/integration/run_backup.py (hunk 2 above)
Comment (atefehmohseni) [reply test/integration/run_backup.py:166]: yes.

File: test/integration/run_backup.py (hunk 8 above)
Comment (atefehmohseni) [reply test/integration/run_backup.py:181]: I see. `raise` will terminate the flow of the program and I dont think it will hit `finally` in this case.

File: test/integration/test_w_valgrind.py (hunk 10 above)
Comment (atefehmohseni) [reply test/integration/test_w_valgrind.py]: Oops, it looks aligned on my IDE (VS code). will fix it directly from Github.

File: val.sh (hunk 11 above)
Comment (atefehmohseni) [reply val.sh]: You're right. I changed it for some specific tests and forgot to revert it.

File: test/integration/test_w_valgrind.py (hunk 12 above)
Comment (atefehmohseni) [reply test/integration/test_w_valgrind.py]: done.

File: test/integration/aerospike_servers.py (hunk 5 above)
Comment (atefehmohseni) [reply test/integration/aerospike_servers.py:25]: No it didnt. Previous test used older version of server (v5.7)
